iwyu.cc

```diff
@@ -1664,6 +1664,10 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
                                                                  comment);
       }
     } else {
+      if (const auto* template_spec_type =
+              dyn_cast<TemplateSpecializationType>(Desugar(type))) {
+        this->getDerived().ReportTemplateSpecTypeInternals(template_spec_type);
```

Comment:
I see the comment on line 1657 says "we want to be careful to call *our* ReportDeclUse" -- we can't really make `ReportTemplateSpecTypeInternals` uphold that constraint, can we? I wonder why, more specifically, it wants to avoid any subclass overrides...

Reply:
It is not fully clear to me. It is important for `ReportDeclUse` to call `IwyuBaseAstVisitor::ReportTypeUse` for typedef underlying type, and not to call `InstantiatedTemplateVisitor::ReportTypeUse`, because the latter resugars the underlying type back into the typedef type, and execution goes into an infinite recursion in that case.
But as for `IwyuBaseAstVisitor::ReportTypeUse`, I tried to remove class qualification from its internal `ReportDeclUse` call, have the tests run and figured out that IWYU stops to report template argument type usage if that type is "provided" by template. In fact, there is currently some difference between reportings of type and declaration usages in an instantiated template: type full uses are always reported for non-defaulted type template arguments, and declaration full uses are reported only if the declaration isn't "provided" by template (e.g., if it is placed in the template containing file or in a file directly included in it). I'm not sure whether types should not actually follow the logic for declarations, but maybe IWYU principle forces explicitly written template arguments to be reported in an instantiating file regardless of whether template defining file provides them or not?
Moreover, after aforementioned change, IWYU starts to report "provided" types in the location of a template. It matters when template defining file is included into analysis and, e.g., type declaration is placed _after_ template definition in the same file. But [I'm not sure if reporting a use into the template definition location is actually reasonable](https://github.com/bolshakov-a/include-what-you-use/commit/ef27dfa1f4fe49e9ef7f7d7f5eae0f4b08a83692).
As concerns `ReportTemplateSpecTypeInternals`, I don't expect any problems with infinite resugaring-desugaring or issues with declaration providing, because it should report types _different_ from original template specialization type.

Reply:
I've been looking at this for a good while, and I honestly can't get my head around it, neither your response above, nor the IWYU implementation :).

Avoiding infinite recursion is probably why the cautionary comment is there, so that makes sense.

So based on your analysis, I guess as long as `ReportTemplateSpecTypeInternals` doesn't call `ReportDeclUse` everything will be fine? I guess it's a little hard to maintain that invariant, but if we can put in comments to remind our future selves, that might be the best we can do right now.

Reply:
The worst thing that can happen is some strange reporting of a template specialization type in location of its own definition, I think. I considered that it might be the case for such a code
```cpp
template <typename T>
struct Tpl
{
  typedef int Type;
  typename Tpl<T>::Type i; // senseless qualification, but compilers allow such a code
};
```
But I didn't succeed in obtaining any wrong behavior.
I think that infinite recursion might occur only on typedef underlying type reporting, but the possible recursion loop is already broken at the [line 1591](https://github.com/include-what-you-use/include-what-you-use/pull/1136/files/8c1f66ba57a3d251c01ceb47edca00128239b11d#diff-7ca107c8b836c1765e8b735b74f06833bec48f308a76017886cee1d5736002d9R1591).
Calling subclass' `ReportDeclUse` from this function (`ReportTypeUse`) is quite safe, as tests have shown.

Reply:
Thank you for getting to the bottom of this. I can't think of a nice 100% safe construct, so let's not worry about this potential problem.